src/optimizer/transforms/char-case-insensitive-lowercase-transform.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * The MIT License (MIT)
+ * Copyright (c) 2017-present Dmitry Soshnikov <dmitry.soshnikov@gmail.com>
+ */
+
+'use strict';
+
+/**
+ * Transforms case-insensitive regexp to lowercase
+ *
+ * /AaBbÏ/i -> /aabbï/i
+ */
+module.exports = {
+  _AZClassRanges: null,
+  _hasUFlag: false,
+  init(ast) {
+    this._AZClassRanges = new Set();
+    this._hasUFlag = ast.flags.includes('u');
+  },
+  shouldRun(ast) {
+    return ast.flags.includes('i');
+  },
+  Char(path) {
+    const {node, parent} = path;
+    if (isNaN(node.codePoint)) {
+      return;
+    }
+
+    // Engine support for case-insensitive matching without the u flag
+    // for characters above \u1000 does not seem reliable.
+    if (!this._hasUFlag && node.codePoint >= 0x1000) {
+      return;
+    }
+
+    if (parent.type === 'ClassRange') {
+      // The only class ranges we handle are inside A-Z
+      if (isAZClassRange(parent)) {
+        this._AZClassRanges.add(parent);
+      }
+      if (!this._AZClassRanges.has(parent)) {
+        return;
+      }
+    }
+
+    const lower = node.symbol.toLowerCase();
+    if (lower !== node.symbol) {
+      node.value = displaySymbolAsValue(lower, node);
+      node.symbol = lower;
+      node.codePoint = lower.codePointAt(0);
+    }
+  }
+};
+
+function isAZClassRange(classRange) {
+  const {from, to} = classRange;
+  // A-Z
+  return from.codePoint >= 0x41 && from.codePoint <= 0x5a &&
+    to.codePoint >= 0x41 && to.codePoint <= 0x5a;
+}
+
+function displaySymbolAsValue(symbol, node) {
+  const codePoint = symbol.codePointAt(0);
+  if (node.kind === 'decimal') {
+    return '\\' + codePoint;
+  }
+  if (node.kind === 'oct') {
+    return '\\0' + codePoint.toString(8);
+  }
+  if (node.kind === 'hex') {
+    return '\\x' + codePoint.toString(16);
+  }
+  if (node.kind === 'unicode') {
+    if (node.isSurrogatePair) {
+      let trail = 0xdc00 + (codePoint + 0x35fdc00) % 0x400;
+      let lead = (codePoint + 0x35fdc00 - trail) / 0x400;
```

Comment:
These magic number probably good to define as constants as well, or add some comments (maybe to extract to a helper function with a doc-comment, explaining what each number means).

Reply:
Ok, Google just gave me an easier algorithm, along with an official reference. So even if the numbers look a bit magic, we should be good to go.